Userland/Services/WindowServer/ScreenBackend.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2022, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include "ScreenLayout.h"
+#include <AK/Error.h>
+#include <AK/RefCounted.h>
+#include <AK/Span.h>
+#include <AK/StringView.h>
+#include <sys/ioctl_numbers.h>
+
+namespace WindowServer {
+
+class Screen;
+
+// Handles device interfacing for the screen.
+class ScreenBackend {
+public:
+    virtual ~ScreenBackend() = default;
+
+protected:
+    virtual ErrorOr<void> open() = 0;
+
+    virtual void set_head_buffer(int index) = 0;
+
+    virtual ErrorOr<void> flush_framebuffer_rects(int buffer_index, Span<FBRect const> rects) = 0;
+
+    virtual ErrorOr<void> unmap_framebuffer() = 0;
+    virtual ErrorOr<void> map_framebuffer() = 0;
+
+    virtual ErrorOr<void> set_head_resolution(FBHeadResolution) = 0;
+    virtual ErrorOr<FBHeadProperties> get_head_properties() = 0;
+
+    bool m_can_device_flush_buffers { true };
+    bool m_can_set_head_buffer { false };
+
+    Gfx::ARGB32* m_framebuffer { nullptr };
+    size_t m_size_in_bytes { 0 };
+    size_t m_back_buffer_offset { 0 };
```

Comment:
The ownptr friendness was just for keeping the constructor private, but that never worked for some reason. The other variables are public because I think it's unnecessary to have Screen access them through functions. The reason that I put these in the ScreenBackend is that ScreenBackend sets and manages them. The Screen just accesses and writes to them, while it tells the ScreenBackend to map or unmap a framebuffer, which changes these values in ways that the Screen should not be concerned with.

Reply:
It's perfectly fine if you want to have the screen backend be a "dumb struct" with public members, the thing that has me scratching my head is how you've mixed paradigms with it.

You have an abstract class with data.

You have everything protected except the destructor

The friendship with screen is awkward.

My advice is to decide if you want to have a class that encapsulates some data and methods, or if you want a dumb struct to hold some state for screen. The "performance overhead" for having trivial inline getters and setters is seriously zero, but the non-idiomatic structure of this base class has some real cognitive load on reviewers and figure maintainers. So, either make the members public, or add getters/setters where required. Making them protected and friending Screen is abusing the friend feature of the language and hiding a cleaner possible architecture.